### PR TITLE
feat: Add hooks implementation

### LINKED ISSUE #5: [SDK] implement hooks
Implement hooks as per [spec](https://github.com/open-feature/spec/blob/main/specification/sections/04-hooks.md).

### PR DESCRIPTION
**Pull request by MaximFischuk**
## This PR
This pull request introduces the implementation of Hooks in the OpenFeature SDK following the OpenFeature standard. The changes include the addition of hooks to various stages in the feature flag evaluation process, such as before, after, error, and finally. The implementation provides capabilities to add hooks at the global, client, provider, and invocation levels, which allows for extensive customizability and control over the feature flag lifecycle.

- Adds hooks to support the OpenFeature specification for flag evaluation.
- Implements the capability to register hooks at various levels including global, client, provider, and invocation.
- Introduces logging for lifecycle events using hooks.
- Provides test cases for the new hooks capability to ensure compliance with the OpenFeature specification.

### Related Issues
Fixes #5 

### Notes
- Make sure logging is appropriately configured in the consuming application to take advantage of the logging hooks.
- This implementation supports extensibility for additional hooks in the future by following the standardized hook interface.

### Follow-up Tasks
- Consider implementing example hooks that demonstrate different use cases, like metrics collection or error reporting.
- Evaluate if default hooks should be provided in the SDK for common use cases.

### How to test
1. Build and run the example included in `examples/hooks.rs` to see a working implementation of the hooks in action.
2. Review test cases provided in `src/hooks/mod.rs` to ensure they cover expected behaviors and edge cases.
3. Ensure that the hooks are called in the correct order and can handle errors gracefully.

### CODE REVIEW

File: src/hooks/logging.rs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    }
+    async fn error<'a>(
+        &self,
+        context: &HookContext<'a>,
+        error: &EvaluationError,
+        _: Option<&'a HookHints>,
+    ) {
+        log::error!("Error hook for flag {}: {:?}", context.flag_key, error);
+    }
+    async fn finally<'a>(&self, context: &HookContext<'a>, _: Option<&'a HookHints>) {
```
Comment (beeme1mr) [inline src/hooks/logging.rs:37]:
```suggestion
    async fn finally<'a>(&self, context: &HookContext<'a>, _: &EvaluationDetails<Value>, _: Option<&'a HookHints>) {
```

Hey @MaximFischuk, the `finally` stage is hooks changed slightly in a recent [spec update](https://github.com/open-feature/spec/pull/280). Thanks!

Comment (MaximFischuk) [reply src/hooks/logging.rs:37]: Thank you for the information! Seems like I've missed it, I'll update it soon :)

**Comment by beeme1mr** [comment]
Hey @MaximFischuk, reviews may be slower than usual around the holidays. Hopefully, someone can take a look soon.

I'll do my best to review, but my Rust skills are basically non-existent 😄

File: src/hooks/logging.rs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+pub struct LoggingHook;
+
+#[async_trait::async_trait]
+impl Hook for LoggingHook {
+    async fn before<'a>(
+        &self,
+        context: &HookContext<'a>,
+        _: Option<&'a HookHints>,
+    ) -> Result<Option<EvaluationContext>, EvaluationError> {
+        log::debug!("Before hook for flag {}", context.flag_key);
```
Comment (gruebel) [inline src/hooks/logging.rs:15]: just thinking, wouldn't it be better to raise the log levels, `debug` -> `info` and the `trace` -> `debug`?

File: src/serde_json.rs [hunk 3]
```diff
@@ -20,7 +20,7 @@ fn json_value_to_value(value: &serde_json::Value) -> EvaluationResult<Value> {
     match value {
         serde_json::Value::Bool(value) => Ok(Value::Bool(*value)),
         serde_json::Value::Number(value) if value.is_i64() => {
-            Ok(Value::Int(value.as_i64().unwrap()))
+            Ok(Value::Int(value.as_i64().unwrap())) //TODO: remove unwrap or use expect
```
Comment (gruebel) [inline src/serde_json.rs:23]: can you explain the TODO? how do you wan to remove the `unwrap()`?

**Review by gruebel — COMMENTED** [review]
overall great job 🍻

File: Cargo.toml [hunk 4]
```diff
@@ -2,7 +2,7 @@
 name = "open-feature"
 version = "0.2.4"
 edition = "2021"
-rust-version = "1.71.1" # MSRV
+rust-version = "1.71.1"                                 # MSRV
```
[the remark this answers is not in the document]
Comment (MaximFischuk) [reply Cargo.toml:5]: Yep, TOML formatter did weird thing :)

File: src/api/api.rs [hunk 5]
```diff
@@ -54,6 +56,12 @@ impl OpenFeature {
         self.provider_registry.set_named(name, provider).await;
     }
 
+    /// Add a new hook to the global list of hooks.
+    pub async fn add_hook<T: Hook>(&mut self, hook: T) {
+        let mut lock = self.hooks.get_mut().await; // TODO: Remove async context to avoid deadlock and enforce initialization before usage.
```
[the remark this answers is not in the document]
Comment (MaximFischuk) [reply src/api/api.rs:61]: Some of my notes were accidentally published to PR, sorry, I'll remove them

File: src/api/client.rs [hunk 6]
```diff
@@ -23,20 +25,26 @@ pub struct Client {
     provider_registry: ProviderRegistry,
     evaluation_context: EvaluationContext,
     global_evaluation_context: GlobalEvaluationContext,
+    global_hooks: GlobalHooks,
+
+    hooks: Vec<HookWrapper>,
```
[the remark this answers is not in the document]
Comment (MaximFischuk) [reply src/api/client.rs:30]: Sure :)

File: src/evaluation/context_field_value.rs [hunk 7]
```diff
@@ -11,7 +11,7 @@ pub enum EvaluationContextFieldValue {
     Float(f64),
     String(String),
     DateTime(OffsetDateTime),
-    Struct(Arc<dyn Any + Send + Sync>),
+    Struct(Arc<dyn Any + Send + Sync>), // TODO: better to make structure in similar way as serde_json::Value
```
[the remark this answers is not in the document]
Comment (MaximFischuk) [reply src/evaluation/context_field_value.rs:14]: Exactly! Type information is completely cleared here, so the client cannot restore the initial type without workarounds. My suggestion is to add a recursion Map like it was implemented in serde_json.

File: src/serde_json.rs (hunk 3 above)
Comment (MaximFischuk) [reply src/serde_json.rs:23]: In general, it is better to avoid any potential panic in libraries, so in my opinion, the error could be returned to the caller as EvaluationResult::Err(...), or when any error is impossible then `expect` will make it clearer to anyone who gets an unexpected panic.

File: src/hooks/logging.rs (hunk 2 above)
Comment (MaximFischuk) [reply src/hooks/logging.rs:15]: It's debatable, enterprise clients usually hide `debug` level of logs in production. Flag evaluating could generate extremely many logs in large applications so it may force a client to disable logs specifically for the OpenFeature library because just setting log level `info` is not enough.
But, I don't insist :)

File: src/hooks/logging.rs [hunk 8]
```diff
@@ -0,0 +1,45 @@
+use crate::{EvaluationContext, EvaluationDetails, EvaluationError, Value};
+
+use super::{Hook, HookContext, HookHints};
+
+/// A hook that logs the evaluation lifecycle of a flag.
+pub struct LoggingHook;
```
Comment (jbovet) [inline src/hooks/logging.rs:6]: @MaximFischuk please consider adding a print option to the LoggingHook struct for evaluation context; it's important to print the context in a readable JSON format if it is set [according to the spec](https://github.com/open-feature/spec/blob/main/specification/appendix-a-included-utilities.md#logging-hook)

File: src/hooks/logging.rs [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+        Ok(None)
+    }
+    async fn after<'a>(
+        &self,
+        context: &HookContext<'a>,
+        _: &EvaluationDetails<Value>,
+        _: Option<&'a HookHints>,
+    ) -> Result<(), EvaluationError> {
+        log::debug!("After hook for flag {}", context.flag_key);
```
Comment (jbovet) [inline src/hooks/logging.rs:25]: please consider adding fields (Logged data) to the log messages (serialized, opt-in) according to [the specification](https://github.com/open-feature/spec/blob/main/specification/appendix-a-included-utilities.md#logging-hook) in each Stage.

File: src/hooks/logging.rs (hunk 8 above)
Comment (MaximFischuk) [reply src/hooks/logging.rs:6]: Is it okay if I define compile-time features as options for `logs` to be structured or not? By default, in Rust, we print logs as formatted strings without any structure, therefore to enable *structured* logs, need to enable specific *feature* (`kv`) of `log` crate. This feature is supported by drastically few implementation crates. (eg. `structured_logger`)

File: src/hooks/logging.rs (hunk 2 above)
Comment (beeme1mr) [reply src/hooks/logging.rs:15]: The spec recommends either `debug` or `info`. I'm fine with `debug` here.

https://openfeature.dev/specification/appendix-a#logging-hook

**Review by beeme1mr — APPROVED** [review]
This PR looks good from a spec perspective.

Before this is merged, could you please update the hooks section in the readme?

- https://openfeature.dev/specification/appendix-a#logging-hook
- https://github.com/open-feature/rust-sdk?tab=readme-ov-file#develop-a-hook

Thanks @MaximFischuk!

**Comment by MaximFischuk** [comment]
@beeme1mr @jbovet @gruebel Thank you for your review!

I've extended the implementation of the logging hook to support both text and structured features of the `log` crate. Unfortunately, it's impossible to serialize `EvaluationContext` into JSON for structured logging without changing `EvaluationContextFieldValue` type for a moment, so I've currently left it as is.

@beeme1mr Also, I've updated `README.md` to include documentation about new features.

**Review by beeme1mr — APPROVED** [review]

**Comment by beeme1mr** [comment]
I'll give folks the weekend to review. If there are no concerns, I'll merge early next week. Thanks @MaximFischuk.

**Review by gruebel — APPROVED** [review]
nice work 🍻

**Review by jbovet — APPROVED** [review]
@MaximFischuk well done!
